### PR TITLE
Add streaming data iter for GBW

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-121/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/2/index.html

**Comment by mli** [comment]
Job PR-121/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/13/index.html

**Comment by mli** [comment]
Job PR-121/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/16/index.html

**Comment by mli** [comment]
Job PR-121/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/17/index.html

**Comment by eric-haibin-lin** [comment]
https://github.com/dmlc/gluon-nlp/issues/131

**Comment by mli** [comment]
Job PR-121/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/20/index.html

**Comment by mli** [comment]
Job PR-121/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/21/index.html

**Comment by mli** [comment]
Job PR-121/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/22/index.html

**Comment by mli** [comment]
Job PR-121/23 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/23/index.html

File: gluonnlp/data/language_model.py [hunk 1]
```diff
@@ -251,3 +254,89 @@ def __init__(self, segment='train', skip_empty=True,
...
+            os.makedirs(root)
+        self._root = root
+        self._dir = os.path.join(root, '1-billion-word-language-modeling-benchmark-r13output')
+        self._namespace = 'gluon/dataset/{}'.format(namespace)
+        subdir_name, pattern, data_hash = self._data_file[segment]
+        self._subdir = os.path.join(self._dir, subdir_name)
+        self._file_pattern = os.path.join(self._subdir, pattern)
+        self._data_hash = data_hash
+        self._get_data()
+        sampler = 'random' if segment == 'train' else 'sequential'
```
Comment (cgraywang) [inline gluonnlp/data/language_model.py]: why sequential is needed? In the evaluation, the model should also use random sampler.

File: gluonnlp/data/stream.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        self._bos = bos
+        self._eos = eos
+        self._sampler = sampler
+        self._file_sampler = file_sampler
+
+    def _get_sampler(self, sampler):
+        assert isinstance(sampler, str), 'Expected sampler to be a str, but got %s'%type(sampler)
+        if sampler == 'random':
+            return RandomSampler
+        if sampler == 'sequential':
```
Comment (cgraywang) [inline gluonnlp/data/stream.py]: same, sequential?

File: gluonnlp/data/stream.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            - keep: A batch with less samples than previous batches is returned.
+            - discard: The last batch is discarded if it's smaller than `(seq_len, batch_size)`.
+        """
+        corpus = CorpusStream(self._file_pattern, flatten=False, encoding=self._encoding,
+                              skip_empty=self._skip_empty, sample_splitter=self._sample_splitter,
+                              tokenizer=self._tokenizer, bos=self._bos, eos=self._eos,
+                              sampler=self._sampler, file_sampler=self._file_sampler)
+        return _LanguageModelBPTTStream(corpus, vocab, seq_len, batch_size, last_batch=last_batch)
+
+class _LanguageModelBPTTStream(DataStream):
```
Comment (cgraywang) [inline gluonnlp/data/stream.py]: We probably should provide character level vocab and language model stream as well.

File: gluonnlp/vocab.py [hunk 4]
```diff
@@ -456,3 +463,126 @@ def from_json(json_str):
...
+        A dict mapping each token to its index integer.
+    unknown_token : hashable object or None
+        The representation for any unknown token. In other words, any unknown token will be indexed
+        as the same representation.
+
+    Examples
+    --------
+    >>> x = nlp.data.Counter({'a': 10, 'b': 1, 'c': 2, '<unk>': 3, '<bos>': 1})
+    >>> vocab = nlp.SortedVocab(counter=x)
+    >>> assert len(vocab0) == 7
```
Comment (cgraywang) [inline gluonnlp/vocab.py]: assert len(vocab) == 7?

File: gluonnlp/data/stream.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        self._file_sampler = file_sampler
+        super(LanguageModelStream, self).__init__(self._file_pattern, flatten=True,
+                                                  encoding=self._encoding,
+                                                  skip_empty=self._skip_empty,
+                                                  sample_splitter=self._sample_splitter,
+                                                  tokenizer=self._tokenizer, bos=self._bos,
+                                                  eos=self._eos, sampler=self._sampler,
+                                                  file_sampler=self._file_sampler)
+
+    def bptt_batchify(self, vocab, seq_len, batch_size, last_batch='keep'):
```
Comment (cgraywang) [inline gluonnlp/data/stream.py]: should we provide batchify-only here? Then we allow users to play with the bptt in user-defined get_batch. I know it is not essential for now since the training and inference should use random sampling. But could be a useful feature.

File: gluonnlp/vocab.py [hunk 6]
```diff
@@ -456,3 +463,126 @@ def from_json(json_str):
         vocab._bos_token = vocab_dict.get('bos_token')
         vocab._eos_token = vocab_dict.get('eos_token')
         return vocab
+
+class SortedVocab(Vocab):
```
Comment (cgraywang) [inline gluonnlp/vocab.py]: Do we necessarily need this class? We probably can make the ``sort`` a transformer for the vocab.

```
>>> x = nlp.data.Counter({'a': 10, 'b': 1, 'c': 2, '<unk>': 3, '<bos>': 1})
>>> y = nlp.data.Counter(x.most_common())
>>> vocab = nlp.Vocab(counter=y)
>>> vocab['a']
0

```

File: gluonnlp/data/stream.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+            return RandomSampler
+        if sampler == 'sequential':
+            return SequentialSampler
+        raise ValueError('sampler must be either "random" or "sequential", but got %s'%(sampler))
+
+    def __iter__(self):
+        sampler = self._get_sampler(self._sampler)
+        file_sampler = self._get_sampler(self._file_sampler)
+        # generate file samples
+        files = glob.glob(self._file_pattern)
```
Comment (sxjscience) [inline gluonnlp/data/stream.py]: Should we use `sorted(glob.glob(self._file_pattern))`?

File: gluonnlp/vocab.py (hunk 6 above)
Comment (eric-haibin-lin) [reply gluonnlp/vocab.py]: The problem with the current Vocab is that special / reserved words are always indexed ahead of common words and unk always indexes to 0. SortedVocab breaks all that assumption

File: gluonnlp/vocab.py (hunk 4 above)
Comment (eric-haibin-lin) [reply gluonnlp/vocab.py]: Good catch. will update

File: gluonnlp/data/stream.py (hunk 7 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/stream.py]: good catch. will update

File: gluonnlp/data/language_model.py (hunk 1 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/language_model.py]: i'll email author to figure it out..

File: gluonnlp/data/stream.py (hunk 2 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/stream.py]: As discussed offline, useful for debugging

File: gluonnlp/data/stream.py (hunk 3 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/stream.py]: Yes. We can extend this class in later PRs.

File: gluonnlp/data/stream.py (hunk 5 above)
Comment (eric-haibin-lin) [reply gluonnlp/data/stream.py]: Discussed offline, not applicable to streams

**Comment by mli** [comment]
Job PR-121/30 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/30/index.html

File: gluonnlp/vocab.py (hunk 6 above)
Comment (cgraywang) [reply gluonnlp/vocab.py]: I got that. But what can be the negative impact of that since you know the index of the special tokens? If you can just sort the vocab, using sort as a transform, why not using it?

File: gluonnlp/data/stream.py (hunk 3 above)
Comment (cgraywang) [reply gluonnlp/data/stream.py]: Yes, let's add the character vocab to support the implementation of the whole model zoo.

File: gluonnlp/data/language_model.py (hunk 1 above)
Comment (cgraywang) [reply gluonnlp/data/language_model.py]: Thanks, the first author and second author have different ways of evaluating the model...

**Comment by mli** [comment]
Job PR-121/31 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/31/index.html

File: gluonnlp/data/stream.py (hunk 3 above)
Comment (szha) [reply gluonnlp/data/stream.py]: we already have a flavor of character vocab in word embedding training. also, the current class already can support character level data by switching the sample splitter.

File: gluonnlp/vocab.py [hunk 8]
```diff
@@ -688,3 +695,126 @@ def subwords_to_indices(self, subwords):
             self.fasttext_hash_asbytes(sw) % self.num_subwords
             for sw in subwords
         ]
+
+class SortedVocab(Vocab):
```
Comment (szha) [inline gluonnlp/vocab.py]: #131

**Comment by eric-haibin-lin** [comment]
@szha I've removed changes to vocab class. I'll submit those in a separate PR.

**Comment by mli** [comment]
Job PR-121/32 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/32/index.html

**Comment by mli** [comment]
Job PR-121/35 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/35/index.html

File: docs/api/data.rst [hunk 9]
```diff
@@ -108,6 +121,20 @@ load custom datasets.
...
+DataStream API for streaming and processing common text formats. The following classes can be used or subclassed to
+stream large custom data.
+
+.. autosummary::
+    :nosignatures:
+
+    DataStream
+    SimpleDataStream
+    CorpusStream
+    LanguageModelStream
```
Comment (szha) [inline docs/api/data.rst]: Links for these are broken.

Comment (szha) [reply docs/api/data.rst]: These classes are not showing up on the API doc. Double-check imports?

**Comment by mli** [comment]
Job PR-121/36 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/36/index.html

**Comment by mli** [comment]
Job PR-121/37 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-121/37/index.html
